src/array_array.jl

```diff
@@ -13,23 +13,39 @@ Compute the primal value `y = f(x)` and the Jacobian matrix `jac = ∂f(x)` of a
 $JAC_NOTES
 """
 function value_and_jacobian!(
-    jac::AbstractMatrix, backend::AbstractBackend, f, x::AbstractArray
+    jac::AbstractMatrix, backend::AbstractADType, f, x::AbstractArray
 )
-    y = f(x)
+    return value_and_jacobian!(Val{:fallback}(), jac, backend, f, x)
+end
+
+function value_and_jacobian!(
+    implem::Val{:fallback},
+    jac::AbstractMatrix,
+    backend::AbstractADType,
+    f,
+    x::AbstractArray,
+)
+    return value_and_jacobian!(implem, autodiff_mode(backend), jac, backend, f, x)
+end
+
+function check_jac(jac::AbstractMatrix, x::AbstractArray, y::AbstractArray)
     nx, ny = length(x), length(y)
     size(jac) != (ny, nx) && throw(
         DimensionMismatch("Size of Jacobian buffer doesn't match expected size ($ny, $nx)"),
     )
-    return _value_and_jacobian!(jac, backend, f, x, y)
+    return nothing
 end
 
-function _value_and_jacobian!(
+function value_and_jacobian!(
+    ::Val{:fallback},
+    ::Val{:forward},
```

Comment:
Is there a way to somehow combine this with `extras` we plan on adding? Maybe we could have a parametric struct `Extras{Mode,Implem}`.

Reply:
This `Extras` struct could wrap other `extras` arguments. We could have default constructor for every backend that sets the mode and implementation.

Something like
```Julia
function value_and_jacobian!(jac, backend, f, x, extras)
    value_and_jacobian!(jac, backend, f, x, Extras(backend, extras)
end
```

This way we could avoid boiler-plate code in all other methods besides the lowest-level `value_and_X!` functions by just forwarding `extras`.

Reply:
I don't really like this option because to me the extras should be purely backend specific, and set by the user or by our own "prepare_gradient" methods.
On the other hand, the fallback mechanism is mostly for our internal use and benchmarking, I don't think users should even know about it. After all, if we find that something performs better with our fallbacks, the goal would be to contribute it to the AD package instead of leaving it here.

Reply:
Finally, note that your approach seems to add one method for each function, so it's mostly the same quantity of "boilerplate" as mine. The only difference is for the utilities where we have to differentiate between forward and reverse mode, but then again you would need to do that too

Reply:
However I do think if we keep my approach we should replace value types with custom structs in the trait mechanism, to avoid accidental piracy